src/223/main/kotlin/org/rust/ide/hints/codeVision/RsImplementationsCodeVisionProvider.kt

```diff
@@ -0,0 +1,96 @@
+import com.intellij.openapi.util.registry.Registry
+import com.intellij.openapi.util.text.StringUtil
+import com.intellij.psi.NavigatablePsiElement
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import org.rust.RsBundle
+import org.rust.lang.core.psi.RsFile
+import org.rust.lang.core.psi.RsTraitItem
+import org.rust.lang.core.psi.ext.*
+import org.rust.openapiext.isUnitTestMode
+import java.awt.event.MouseEvent
+
+class RsImplementationsCodeVisionProvider : InheritorsCodeVisionProvider() {
+    override fun acceptsFile(file: PsiFile): Boolean = file is RsFile
+
+    override fun acceptsElement(element: PsiElement): Boolean {
+        if (!isUnitTestMode && !Registry.`is`("org.rust.code.vision.implementation", false)) return false
+
+        return when (element) {
+            is RsTraitItem -> true
+            is RsAbstractable -> element.owner is RsAbstractableOwner.Trait
+            else -> false
+        }
+    }
+
+    override fun getHint(element: PsiElement, file: PsiFile): String? {
+        val implementationCount = getImplementations(element).size
+        if (implementationCount == 0) return null
+
+        return when (element) {
+            is RsTraitItem -> {
+                return RsBundle.message("rust.code.vision.implementation.hint", implementationCount)
+            }
+            is RsAbstractable -> {
+                return if (element.isAbstract) {
+                    RsBundle.message("rust.code.vision.implementation.hint", implementationCount)
+                } else {
+                    RsBundle.message("rust.code.vision.overrides.hint", implementationCount)
+                }
+            }
+            else -> null
+        }
+    }
+
+    override fun handleClick(editor: Editor, element: PsiElement, event: MouseEvent?) {
+        // copied from ImplsNavigationGutterIconRenderer::navigateToItems
+        if (event == null) return
+
+        val namedElement = element as? RsNamedElement ?: return
+        val elementName = namedElement.name ?: return
+
+        val targets = getImplementations(element).filterIsInstance<NavigatablePsiElement>().toTypedArray()
+        if (targets.isEmpty()) return
+
+        val escapedName = StringUtil.escapeXmlEntities(elementName)
+
+        PsiElementListNavigator.openTargets(
+            event,
+            targets,
+            CodeInsightBundle.message("goto.implementation.chooserTitle", escapedName, targets.size, ""),
+            CodeInsightBundle.message("goto.implementation.findUsages.title", escapedName, targets.size),
+            DefaultPsiElementCellRenderer()
+        )
+    }
+
+    private fun getImplementations(element: PsiElement): List<PsiElement> {
+        val query = when (element) {
+            is RsTraitItem -> element.searchForImplementations()
+            is RsAbstractable -> element.searchForImplementations()
+            else -> return emptyList()
+        }
+        return query.findAll().toList()
```

Comment:
If the name is too popular, this query is going to run forever and will be recalculated on every keystroke. For such names it is not worth to do the search, just skip. Also, you don't need to actually store elements, just the size is important.
Some kind of cache is also required, because inheritors depends only on the structure of declarations.

Reply:
@sirgl thank you for reviewing this!